crates/typst-library/src/pdf/embed.rs

```diff
@@ -0,0 +1,216 @@
+        #[named]
+        name: Option<Option<EcoString>>,
+        /// A description for the attached file
+        #[named]
+        description: Option<Option<EcoString>>,
+        /// The mime-type of the embedded file
+        #[named]
+        mime_type: Option<Option<EcoString>>,
+        /// The mime-type of the embedded file
+        #[named]
+        relationship: Option<Option<EmbeddedFileRelationship>>,
+    ) -> StrResult<Content> {
+        let mut elem = EmbedElem::new(path, data);
+        if let Some(name) = name {
+            elem.push_name(name);
+        }
+        if let Some(description) = description {
+            elem.push_description(description);
+        }
+        if let Some(mime_type) = mime_type {
+            elem.push_mime_type(mime_type);
+        }
+        if let Some(relationship) = relationship {
+            elem.push_relationship(relationship);
+        }
+
+        Ok(elem.pack().spanned(span))
+    }
+}
+
+impl LocalName for Packed<EmbedElem> {
+    const KEY: &'static str = "embedding";
+}
+
+impl Show for Packed<EmbedElem> {
+    fn show(&self, _: &mut Engine, _: StyleChain) -> SourceResult<Content> {
+        Ok(Content::empty())
+    }
+}
+
+/// The internal representation of a file embedding
+#[derive(Hash)]
+pub struct Repr {
+    /// The raw file data.
+    data: Bytes,
+    /// Path of this embedding
+    path: EcoString,
+    /// Name of this embedding
+    name: EcoString,
+    /// Name of this embedding
+    description: Option<EcoString>,
+    /// Name of this embedding
+    mime_type: Option<EcoString>,
+    /// Name of this embedding
+    relationship: Option<EmbeddedFileRelationship>,
+}
+
+/// The relationship of an embedded file with the relevant document content
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Cast)]
+pub enum EmbeddedFileRelationship {
+    /// The embedded file is the original source material of the document content
+    Source,
+    /// The embedded file represents information used to derive a visual presentation – such
+    /// as for a table or a graph.
+    Data,
+    /// The embedded file is an alternative representation of document content
+    Alternative,
+    /// The embedded file is a supplemental representation of document content
+    Supplement,
+    /// The embedded file is encrypted and should be displayed to the user if
+    /// the PDF processor has the cryptographic filter needed to
+    /// decrypt the document.
+    EncryptedPayload,
+    /// The embedded file is data associated with an AcroForm
+    FormData,
+    /// The embedded file is a schema definition
+    Schema,
+    /// The embedded file has an unknown relationship to the document or the relationship cannot be
+    /// described by the other variants
+    Unspecified,
+}
+
+/// A loaded file to be embedded.
+///
+/// Values of this type are cheap to clone and hash.
+#[derive(Clone, Hash, Eq, PartialEq)]
+pub struct Embed(Arc<LazyHash<Repr>>);
```

Comment:
`from_element` has a little bit of logic to fallback on using the file path as the name.

I tried using the Packed<EmbedElem>, but cannot access the optional fields because they are private.

Reply:
You should be able to access the private fields via accessor function, like mentinoed above: `element.description(StyleChain::default())`.